File: packages/wasmvm/wasmclient/src/isc/offledgerrequest.rs
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use crate::keypair;
use wasmlib::*;

//TODO generalize this trait
pub trait OffLedgerRequest<'a> {
    fn new(
        chain_id: &ScChainID,
        contract: &ScHname,
        entry_point: &ScHname,
        params: &ScDict,
        signature_scheme: Option<&'a OffLedgerSignatureScheme>,
        nonce: u64,
    ) -> Self;
    fn with_nonce(&mut self, nonce: u64) -> &Self;
    fn with_gas_budget(&mut self, gas_budget: u64) -> &Self;
    fn with_allowance(&mut self, allowance: &ScAssets) -> &Self;
    fn sign(&mut self, key: &keypair::KeyPair) -> &Self;
}

#[derive(Clone)]
pub struct OffLedgerRequestData<'a> {
    chain_id: ScChainID,
    contract: ScHname,
    entry_point: ScHname,
    params: ScDict,
    signature_scheme: Option<&'a OffLedgerSignatureScheme>, // None if unsigned
    nonce: u64,
    allowance: ScAssets,
    gas_budget: u64,
}

pub struct OffLedgerSignatureScheme {
    key_pair: keypair::KeyPair,
    signature: Vec<u8>,
}

impl OffLedgerSignatureScheme {
    pub fn clone(&self) -> Self {
        todo!()
    }
}

impl<'a> OffLedgerRequest<'a> for OffLedgerRequestData<'a> {
    fn new(
        chain_id: &ScChainID,
        contract: &ScHname,
        entry_point: &ScHname,
        params: &ScDict,
        signature_scheme: Option<&'a OffLedgerSignatureScheme>,
        nonce: u64,
    ) -> Self {
        return OffLedgerRequestData {
            chain_id: chain_id.clone(),
            contract: contract.clone(),
            entry_point: entry_point.clone(),
            params: params.clone(),
            signature_scheme: match signature_scheme {
                Some(signature_scheme_val) => Some(signature_scheme_val.to_owned()),
                None => None,
            },
            nonce: nonce,
            allowance: ScAssets::new(&Vec::new()),
            gas_budget: super::gas::MAX_GAS_PER_REQUEST,
        };
    }
    fn with_nonce(&mut self, nonce: u64) -> &Self {
        self.nonce = nonce;
        return self;
    }
    fn with_gas_budget(&mut self, gas_budget: u64) -> &Self {
        self.gas_budget = gas_budget;
        return self;
    }
    fn with_allowance(&mut self, allowance: &ScAssets) -> &Self {
        self.allowance = allowance.clone();
        return self;
    }
    fn sign(&mut self, _key: &keypair::KeyPair) -> &Self {
        todo!()
    }
}

impl<'a> OffLedgerRequestData<'a> {
    pub fn id(&self) -> ScRequestID {
        todo!()
    }
    pub fn essence(&self) -> Vec<u8> {
        let mut data: Vec<u8> = vec![1];
        data.append(self.chain_id.to_bytes().as_mut());
        data.append(self.contract.to_bytes().as_mut());
        data.append(self.entry_point.to_bytes().as_mut());
        data.append(self.params.to_bytes().as_mut());
        data.append(wasmlib::uint64_to_bytes(self.nonce).as_mut());
        data.append(wasmlib::uint64_to_bytes(self.gas_budget).as_mut());
        let scheme = match self.signature_scheme {
            Some(val) => val.clone(),
            None => {
                panic!("none")
            }
        };
        let mut public_key = scheme.key_pair.public_key.to_bytes().to_vec();
        data.push(public_key.len() as u8);
        data.append(&mut public_key);
        data.append(self.allowance.to_bytes().as_mut());
        return data;
    }
    pub fn to_bytes(&self) -> Vec<u8> {
        todo!()
    }
}

File: packages/wasmvm/wasmclient/ts/wasmclient/lib/isc/offledgerrequest.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {Blake2b} from '@iota/crypto.js';
import * as wasmlib from 'wasmlib';
import {concat} from 'wasmlib';
import {KeyPair} from './keypair';

export class OffLedgerSignatureScheme {
    keyPair: KeyPair;
    signature: Uint8Array;

    public constructor(keyPair: KeyPair) {
        this.keyPair = keyPair;
        this.signature = new Uint8Array(0);
    }
}

export class OffLedgerRequest {
    chainID: wasmlib.ScChainID;
    contract: wasmlib.ScHname;
    entryPoint: wasmlib.ScHname;
    params: Uint8Array;
    signatureScheme: OffLedgerSignatureScheme = new OffLedgerSignatureScheme(new KeyPair(new Uint8Array(0)));
    nonce: u64;
    allowance: wasmlib.ScAssets = new wasmlib.ScAssets(new Uint8Array(0));
    gasBudget: u64 = 2n ** 64n - 1n;

    public constructor(chainID: wasmlib.ScChainID, contract: wasmlib.ScHname, entryPoint: wasmlib.ScHname, params: Uint8Array, nonce: u64) {
        this.chainID = chainID;
        this.contract = contract;
        this.entryPoint = entryPoint;
        this.params = params;
        this.nonce = nonce;
    }

    public bytes(): Uint8Array {
        const sig = this.signatureScheme.signature;
        const data = concat(this.essence(), wasmlib.uint16ToBytes(sig.length as u16));
        return concat(data, sig);
    }

    public essence(): Uint8Array {
        const oneByte = new Uint8Array(1);
        oneByte[0] = 1; // requestKindTagOffLedgerISC
        let data = concat(oneByte, this.chainID.toBytes());
        data = concat(data, this.contract.toBytes());
        data = concat(data, this.entryPoint.toBytes());
        data = concat(data, this.params);
        data = concat(data, wasmlib.uint64ToBytes(this.nonce));
        data = concat(data, wasmlib.uint64ToBytes(this.gasBudget));
        const pubKey = this.signatureScheme.keyPair.publicKey;
        oneByte[0] = pubKey.length as u8;
        data = concat(data, oneByte);
        data = concat(data, pubKey);
        data = concat(data, this.allowance.toBytes());
        return data;
    }

    public ID(): wasmlib.ScRequestID {
        // req id is hash of req bytes with output index zero
        const hash = Blake2b.sum256(this.bytes());
        const reqId = new wasmlib.ScRequestID();
        reqId.id.set(hash, 0);
        return reqId;
    }

    public sign(keyPair: KeyPair): OffLedgerRequest {
        const req = new OffLedgerRequest(this.chainID, this.contract, this.entryPoint, this.params, this.nonce);
        req.signatureScheme = new OffLedgerSignatureScheme(keyPair);
        req.signatureScheme.signature = keyPair.sign(req.essence());
        return req;
    }

    public withAllowance(allowance: wasmlib.ScAssets): void {
        this.allowance = allowance;
    }
}
